sdk/core/core-http/lib/policies/atomSerializationPolicy.ts

```diff
@@ -0,0 +1,185 @@
+import { isString, byteLength } from "../util/utils";
+import { ResourceSerializer } from "../resourceSerializer";
+
+/**
+ * Create a new serialization RequestPolicyCreator that will serialize/deserialize
+ * HTTP request bodies as they pass through the HTTP pipeline.
+ */
+export function atomSerializationPolicy(): RequestPolicyFactory {
+  return {
+    create: (nextPolicy: RequestPolicy, options: RequestPolicyOptions) => {
+      return new AtomSerializationPolicy(nextPolicy, options);
+    }
+  };
+}
+
+/**
+ * A RequestPolicy that will
+ *  - serialize HTTP requests with input in JSON to ATOM based XML requests, and
+ *  - deserialize the ATOM based XML responses as they pass through the HTTP pipeline.
+ */
+export class AtomSerializationPolicy extends BaseRequestPolicy {
+  constructor(nextPolicy: RequestPolicy, options: RequestPolicyOptions) {
+    super(nextPolicy, options);
+  }
+
+  public async sendRequest(request: WebResource): Promise<HttpOperationResponse> {
+    let shouldParseResponse = false;
+    let serializer: ResourceSerializer;
+    if (request.atomXmlOperationSpec) {
+      serializer = request.atomXmlOperationSpec.serializer;
+      shouldParseResponse = request.atomXmlOperationSpec.shouldParseResponse;
+      if (request.body) {
+        request.body = serializer.serialize(JSON.parse(request.body));
+      }
+    }
+
+    return this._nextPolicy.sendRequest(request).then((response: HttpOperationResponse) => {
+      const parsedResponse: HttpOperationResponse = this.parseResponse(response);
+
+      // Construct response with 'result' to be backward compatibile
+      const responseInCustomJson: any = {
+        error: parsedResponse.errorBody,
+        response: parsedResponse.parsedBody,
+        result: undefined
+      };
+      responseInCustomJson.result =
+        shouldParseResponse && serializer
+          ? serializer.parse(responseInCustomJson.response)
+          : undefined;
+
+      response.parsedBody = responseInCustomJson;
+      return response;
+    });
+  }
+
+  /**
+   * Process the response.
+   * @ignore
+   *
+   * @param {WebResource} webResource  The web resource that made the request.
+   * @param {Response}    response     The response object.
+   * @return The normalized responseObject.
+   */
+  private parseResponse(response: HttpOperationResponse): HttpOperationResponse {
+    const parsedResponse: HttpOperationResponse = this._parseXmlResponseToJson(response);
+
+    if (response.status >= 200 && response.status < 300 && response.errorBody == undefined) {
+      return response;
+    }
+
+    const HttpResponseCodes: any = Constants.HttpResponseCodes;
+
+    if (parsedResponse.errorBody == undefined) {
+      const code = Object.keys(HttpResponseCodes).filter(function(name: any): any {
+        if (HttpResponseCodes[name] === response.status) {
+          return name;
+        }
+      });
+
+      parsedResponse.errorBody = { error: { code: code[0] } };
```

Comment:
Agree with overall simplification - also, since the original implementation was iterating over an existing set of items, this was guaranteed to be not empty.